core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -283,9 +293,14 @@ private[transaction] class TransactionMetadata(val producerId: Long,
             txnStartTimestamp = newMetadata.txnStartTimestamp
             topicPartitions.clear()
           }
+        case Dead => // from expire transactionalId
+          if(producerEpoch != newMetadata.producerEpoch)
+            throw new IllegalStateException("Completing transaction state transition failed due to unexpected metadata")
```

Comment:
In these and other messages, we should include all the transaction metadata, including transcationalId, producerId, producerEpoch, and the precise from and to state transition which is invalid.